control-plane/config-entries/controllersv2/configentry_controller.go

```diff
@@ -0,0 +1,304 @@
+		//
+		// In terms of performance, Consul servers can handle tens of thousands
+		// of writes per second, so retrying at max every 5s isn't an issue and
+		// provides a better UX.
+		RateLimiter: workqueue.NewMaxOfRateLimiter(
+			workqueue.NewItemExponentialFailureRateLimiter(200*time.Millisecond, 5*time.Second),
+			// 10 qps, 100 bucket size.  This is only for retry speed and its only the overall factor (not per item)
+			&workqueue.BucketRateLimiter{Limiter: rate.NewLimiter(rate.Limit(10), 100)},
+		),
+	}
+
+	return ctrl.NewControllerManagedBy(mgr).
+		For(resource).
+		WithOptions(options).
+		Complete(reconciler)
+}
+
+func (r *ConfigEntryController) consulNamespace(namespace string) string {
+	ns := namespaces.ConsulNamespace(
+		namespace,
+		r.EnableConsulNamespaces,
+		r.ConsulDestinationNamespace,
+		r.EnableNSMirroring,
+		r.NSMirroringPrefix,
+	)
+
+	// TODO: remove this if and when the default namespace of resources is no longer required to be set explicitly.
+	if ns == "" {
+		ns = constants.DefaultConsulNS
+	}
+	return ns
+}
+
+func (r *ConfigEntryController) syncFailed(ctx context.Context, logger logr.Logger, updater Controller, configEntry common.ConfigEntryV2Resource, errType string, err error) (ctrl.Result, error) {
+	configEntry.SetSyncedCondition(corev1.ConditionFalse, errType, err.Error())
+	if updateErr := updater.UpdateStatus(ctx, configEntry); updateErr != nil {
+		// Log the original error here because we are returning the updateErr.
+		// Otherwise the original error would be lost.
+		logger.Error(err, "sync failed")
+		return ctrl.Result{}, updateErr
+	}
+	return ctrl.Result{}, err
+}
+
+func (r *ConfigEntryController) syncSuccessful(ctx context.Context, updater Controller, configEntry common.ConfigEntryV2Resource) (ctrl.Result, error) {
+	configEntry.SetSyncedCondition(corev1.ConditionTrue, "", "")
+	timeNow := metav1.NewTime(time.Now())
+	configEntry.SetLastSyncedTime(&timeNow)
+	return ctrl.Result{}, updater.UpdateStatus(ctx, configEntry)
+}
+
+func (r *ConfigEntryController) syncUnknown(ctx context.Context, updater Controller, configEntry common.ConfigEntryV2Resource) error {
+	configEntry.SetSyncedCondition(corev1.ConditionUnknown, "", "")
+	return updater.Update(ctx, configEntry)
+}
+
+func (r *ConfigEntryController) syncUnknownWithError(ctx context.Context,
+	logger logr.Logger,
+	updater Controller,
+	configEntry common.ConfigEntryV2Resource,
+	errType string,
+	err error) (ctrl.Result, error) {
+
+	configEntry.SetSyncedCondition(corev1.ConditionUnknown, errType, err.Error())
+	if updateErr := updater.UpdateStatus(ctx, configEntry); updateErr != nil {
+		// Log the original error here because we are returning the updateErr.
+		// Otherwise the original error would be lost.
+		logger.Error(err, "sync status unknown")
+		return ctrl.Result{}, updateErr
+	}
+	return ctrl.Result{}, err
+}
+
+func isNotFoundErr(err error) bool {
+	return err != nil && strings.Contains(err.Error(), "404")
+}
```

Comment:
I'm sure you're still working on this, but it will be the GRPC not found code: https://github.com/hashicorp/consul-k8s/blob/main/control-plane/connect-inject/controllers/pod/pod_controller_test.go#L1516